File: plugins/plugin-line/rust/src/actions/send_flex_message.rs
//! Send flex message action for the LINE plugin.

use crate::service::LineService;
use crate::types::*;
use serde::{Deserialize, Serialize};
use tracing::debug;

/// Parameters for sending a LINE flex message
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendFlexMessageParams {
    /// Alternative text for notifications
    pub alt_text: String,
    /// Card title
    pub title: String,
    /// Card body
    pub body: String,
    /// Target user/group/room ID
    pub to: String,
}

/// Result of the send flex message action
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SendFlexMessageResult {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub to: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Create a simple info card bubble
fn create_info_bubble(title: &str, body: &str) -> serde_json::Value {
    serde_json::json!({
        "type": "bubble",
        "body": {
            "type": "box",
            "layout": "vertical",
            "contents": [
                {
                    "type": "text",
                    "text": title,
                    "weight": "bold",
                    "size": "xl",
                    "wrap": true
                },
                {
                    "type": "text",
                    "text": body,
                    "margin": "md",
                    "wrap": true
                }
            ]
        }
    })
}

/// Execute the send flex message action
pub async fn execute_send_flex_message(
    service: &LineService,
    params: SendFlexMessageParams,
) -> SendFlexMessageResult {
    // Check connection
    if !service.is_connected().await {
        return SendFlexMessageResult {
            success: false,
            to: None,
            message_id: None,
            error: Some("LINE service not connected".to_string()),
        };
    }

    // Validate target
    let target = match normalize_line_target(&params.to) {
        Some(t) if is_valid_line_id(&t) => t,
        _ => {
            return SendFlexMessageResult {
                success: false,
                to: None,
                message_id: None,
                error: Some("Invalid target ID".to_string()),
            }
        }
    };

    // Create flex message
    let flex_message = LineFlexMessage {
        alt_text: params.alt_text[..params.alt_text.len().min(400)].to_string(),
        contents: create_info_bubble(&params.title, &params.body),
    };

    // Send message
    let result = service.send_flex_message(&target, flex_message).await;

    if !result.success {
        return SendFlexMessageResult {
            success: false,
            to: Some(target),
            message_id: None,
            error: result.error,
        };
    }

    debug!("Sent LINE flex message to {}", target);

    SendFlexMessageResult {
        success: true,
        to: Some(target),
        message_id: result.message_id,
        error: None,
    }
}

/// Action metadata
pub const SEND_FLEX_MESSAGE_ACTION_NAME: &str = "LINE_SEND_FLEX_MESSAGE";
pub const SEND_FLEX_MESSAGE_ACTION_DESCRIPTION: &str = "Send a rich flex message/card via LINE";
pub const SEND_FLEX_MESSAGE_ACTION_SIMILES: &[&str] = &[
    "SEND_LINE_CARD",
    "LINE_FLEX",
    "LINE_CARD",
    "SEND_LINE_FLEX",
];

File: plugins/plugin-line/typescript/src/actions/sendFlexMessage.ts
/**
 * Send flex message action for the LINE plugin.
 */

import type {
	Action,
	ActionResult,
	HandlerCallback,
	IAgentRuntime,
	Memory,
	State,
} from "@elizaos/core";
import {
	composePromptFromState,
	logger,
	ModelType,
	parseJSONObjectFromText,
} from "@elizaos/core";
import type { LineService } from "../service.js";
import {
	isValidLineId,
	LINE_SERVICE_NAME,
	type LineFlexMessage,
	normalizeLineTarget,
} from "../types.js";

const SEND_FLEX_TEMPLATE = `# Task: Extract LINE Flex message parameters

Based on the conversation, determine the flex message content to send.

Recent conversation:
{{recentMessages}}

Extract the following:
1. altText: Alternative text for notifications (short summary)
2. title: Card title
3. body: Card body text
4. to: The target user/group/room ID (or "current" to reply to the current chat)
5. cardType: Type of card (info, image, action, list)

Respond with a JSON object:
\`\`\`json
{
  "altText": "notification text",
  "title": "Card Title",
  "body": "Card body text",
  "to": "target ID or 'current'",
  "cardType": "info"
}
\`\`\`
`;

interface FlexMessageParams {
	altText: string;
	title: string;
	body: string;
	to: string;
	cardType: string;
}

/**
 * Create a simple info card bubble
 */
function createInfoBubble(
	title: string,
	body: string,
): { type: string; [key: string]: unknown } {
	return {
		type: "bubble",
		body: {
			type: "box",
			layout: "vertical",
			contents: [
				{
					type: "text",
					text: title,
					weight: "bold",
					size: "xl",
					wrap: true,
				},
				{
					type: "text",
					text: body,
					margin: "md",
					wrap: true,
				},
			],
		},
	};
}

export const sendFlexMessage: Action = {
	name: "LINE_SEND_FLEX_MESSAGE",
	similes: ["SEND_LINE_CARD", "LINE_FLEX", "LINE_CARD", "SEND_LINE_FLEX"],
	description: "Send a rich flex message/card via LINE",

	validate: async (
		_runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> => {
		return message.content.source === "line";
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state: State | undefined,
		_options?: Record<string, unknown>,
		callback?: HandlerCallback,
	): Promise<ActionResult> => {
		const lineService = runtime.getService(LINE_SERVICE_NAME) as unknown as
			| LineService
			| undefined;

		if (!lineService || !lineService.isConnected()) {
			if (callback) {
				callback({ text: "LINE service is not available.", source: "line" });
			}
			return { success: false, error: "LINE service not available" };
		}

		const currentState = state ?? (await runtime.composeState(message));

		const prompt = composePromptFromState({
			template: SEND_FLEX_TEMPLATE,
			state: currentState,
		});

		let flexInfo: FlexMessageParams | null = null;

		for (let attempt = 0; attempt < 3; attempt++) {
			const response = await runtime.useModel(ModelType.TEXT_SMALL, {
				prompt,
			});

			const parsed = parseJSONObjectFromText(response);
			if (parsed?.title && parsed?.body) {
				flexInfo = {
					altText: String(parsed.altText || `${parsed.title}: ${parsed.body}`),
					title: String(parsed.title),
					body: String(parsed.body),
					to: String(parsed.to || "current"),
					cardType: String(parsed.cardType || "info"),
				};
				break;
			}
		}

		if (!flexInfo || !flexInfo.title) {
			if (callback) {
				callback({
					text: "I couldn't understand the flex message content. Please try again.",
					source: "line",
				});
			}
			return {
				success: false,
				error: "Could not extract flex message parameters",
			};
		}

		// Determine target
		let targetId: string | undefined;

		if (flexInfo.to && flexInfo.to !== "current") {
			const normalized = normalizeLineTarget(flexInfo.to);
			if (normalized && isValidLineId(normalized)) {
				targetId = normalized;
			}
		}

		// Fall back to current chat
		if (!targetId) {
			const stateData = (currentState.data || {}) as Record<string, unknown>;
			targetId =
				(stateData.groupId as string) ||
				(stateData.roomId as string) ||
				(stateData.userId as string);
		}

		if (!targetId) {
			if (callback) {
				callback({
					text: "I couldn't determine where to send the message. Please specify a target.",
					source: "line",
				});
			}
			return { success: false, error: "Could not determine target" };
		}

		// Create flex message
		const flexMessage: LineFlexMessage = {
			altText: flexInfo.altText.slice(0, 400),
			contents: createInfoBubble(flexInfo.title, flexInfo.body),
		};

		// Send message
		const result = await lineService.sendFlexMessage(targetId, flexMessage);

		if (!result.success) {
			if (callback) {
				callback({
					text: `Failed to send flex message: ${result.error}`,
					source: "line",
				});
			}
			return { success: false, error: result.error };
		}

		logger.debug(`Sent LINE flex message to ${targetId}`);

		if (callback) {
			callback({
				text: "Card message sent successfully.",
				source: message.content.source as string,
			});
		}

		return {
			success: true,
			text: "Card message sent successfully",
		};
	},

	examples: [
		[
			{
				name: "{{user1}}",
				content: {
					text: "Send them an info card with title 'Update' and body 'New features are available'",
				},
			},
			{
				name: "{{agent}}",
				content: {
					text: "I'll send that as a card message.",
					actions: ["LINE_SEND_FLEX_MESSAGE"],
				},
			},
		],
	],
};
